cli/lib/stats.ts

```diff
@@ -0,0 +1,92 @@
+import { isSameDay, isSameMonth, isSameWeek } from 'date-fns';
+import fetch from 'node-fetch';
+import { readAppdata, saveAppdata } from './appdata';
+
+export enum StatsGroup {
+	STUDIO_CLI_USAGE_UNIQUE = 'studio-cli-usage-unique',
+}
+
+export enum StatsMetric {
+	SUCCESS = 'success',
+	FAILURE = 'failure',
+}
+
+export type AggregateInterval = 'daily' | 'weekly' | 'monthly';
+
+// Bumps a stat if it hasn't been bumped within the current aggregate interval.
+// This allows us to approximate a 1-count-per-user stat without recording which
+// user the event came from (bump stats are anonymous).
+//
+// We don't want to block the thread to record the stat, so this function doesn't
+// await promises before returning.
+export function bumpAggregatedUniqueStat(
+	group: StatsGroup,
+	stat: StatsMetric,
+	aggregateBy: AggregateInterval,
+	bumpInDev = false
+) {
+	getLastBump( group, stat )
+		.then( ( lastBump ) => {
+			if ( lastBump === null ) {
+				// Bump the stat the first time it's seen
+				bumpStat( group, stat, bumpInDev );
+				return true;
+			}
+
+			const now = Date.now();
+
+			if ( aggregateBy === 'daily' && isSameDay( lastBump, now ) ) {
+				return false;
+			}
+			if ( aggregateBy === 'weekly' && isSameWeek( lastBump, now ) ) {
+				return false;
+			}
+			if ( aggregateBy === 'monthly' && isSameMonth( lastBump, now ) ) {
+				return false;
+			}
+
+			// Bump the stat for subsequent occurrences within the time interval
+			return bumpStat( group, stat, bumpInDev );
+		} )
+		.then( ( didBump ) => {
+			if ( didBump ) {
+				updateLastBump( group, stat );
+			}
+		} );
+}
+
+// Returns true if we attempted to bump the stat
+export function bumpStat( group: StatsGroup, stat: StatsMetric, bumpInDev = false ) {
+	if ( process.env.E2E || ( process.env.NODE_ENV === 'development' && ! bumpInDev ) ) {
```

Comment:
```suggestion
	if ( process.env.NODE_ENV === 'development' && ! bumpInDev ) {
```
I guess we can probably remove the E2E env var check here?